src/app/marketplace/services/awaiting-finalisation.service.ts

```diff
@@ -0,0 +1,170 @@
+  private _purchaseCancelSubscription?: Subscription;
+  private _purchaseFinaliseSubscription?: Subscription;
+  private _fetchSubscription?: Subscription;
+
+  private _productsSubject = new BehaviorSubject<DataProduct[]>([]);
+
+  constructor(
+    private _walletService: WalletService,
+    private _dataProductService: DataProductService,
+    private _dataProductListService: DataProductListService
+  ) {
+    this._walletSubscription = this._walletService.getWallet().subscribe(wallet => this._onWalletChange(wallet));
+  }
+
+  getProducts(): Observable<DataProduct[]> {
+    return this._productsSubject.asObservable();
+  }
+
+  getProductsValue(): DataProduct[] {
+    return this._productsSubject.getValue();
+  }
+
+  refresh(wallet: Wallet): void {
+    if (!wallet) {
+      this._productsSubject.next([]);
+
+      return;
+    }
+
+    if (this._fetchSubscription) {
+      this._fetchSubscription.unsubscribe();
+    }
+
+    this._fetchSubscription = this._dataProductListService
+      .getDataProducts(getAwaitingFinalisationDataProductsQuery(wallet.address), '', environment.maxNotificationsProductsNumber).pipe(
+        pluck('hits')
+      )
+      .subscribe((result: EsDataProduct[]) => {
+        this._productsSubject.next(result.map(esDataProduct => esDataProduct.source));
+      });
+  }
+
+  addProduct(dataProduct: DataProduct): void {
+    const currentValue = this.getProductsValue();
+    if (!currentValue.find(_dataProduct => _dataProduct.address === dataProduct.address)) {
+      this._productsSubject.next([ ...currentValue, dataProduct ]);
+    }
+  }
+
+  removeProduct(dataProduct: DataProduct): void {
+    this._productsSubject.next(
+      this.getProductsValue()
+        .filter(_dataProduct => _dataProduct.address !== dataProduct.address)
+    );
+  }
+
+  findProduct(dataProductAddress: string): DataProduct {
+    return this.getProductsValue().find(dataProduct => dataProduct.address === dataProductAddress);
+  }
+
+  ngOnDestroy(): void {
+    if (this._walletSubscription) {
+      this._walletSubscription.unsubscribe();
+    }
+
+    if (this._updateSubscription) {
+      this._updateSubscription.unsubscribe();
+    }
+
+    if (this._updateSubscription) {
+      this._purchaseSubscription.unsubscribe();
+    }
+
+    if (this._purchaseCancelSubscription) {
+      this._purchaseCancelSubscription.unsubscribe();
+    }
+
+    if (this._purchaseFinaliseSubscription) {
+      this._purchaseFinaliseSubscription.unsubscribe();
+    }
+
+    if (this._fetchSubscription) {
+      this._fetchSubscription.unsubscribe();
+    }
+
+    this._productsSubject.complete();
+  }
+
+  private _onWalletChange(wallet: Wallet): Wallet {
+    if (this._wallet === wallet) {
+      return;
+    }
+
+    const previousWalletValue = this._wallet;
+    this._wallet = wallet;
+
+    if (!previousWalletValue && wallet) {
+      this._updateSubscription = this._dataProductService.watchForDataProductUpdate(null, DataProductUpdateAction.UPDATE)
```

Comment:
You should unsubscribe previous subscriptions first. Every wallet change creates new subscription